### PR TITLE
gossip: remove InfoDeltaStore

### PR DESCRIPTION
**Pull request by tamird**
Best to review this commit-by-commit.

### CODE REVIEW

**Comment by tamird** [comment]
This addresses lingering TODOs from @thschroeter.

File: gossip/client.go [hunk 1]
```diff
@@ -41,7 +41,6 @@ type client struct {
 	rpcClient   *rpc.Client   // RPC client
 	forwardAddr net.Addr      // Set if disconnected with an alternate addr
 	lastFresh   int64         // Last wall time client received fresh info
-	err         error         // Set if client experienced an error
```
Comment (spencerkimball) [inline gossip/client.go]: Why did you remove this? Seems somewhat useful to be able to log this info from where client'a completion is actually handled as opposed to an undifferentiated info log.

File: gossip/client.go [hunk 2]
```diff
@@ -60,23 +59,29 @@ func newClient(addr net.Addr) *client {
...
+			if err != nil {
+				log.Infof("gossip client to %s: %s", c.addr, err)
+			}
+		}()
+
 		c.rpcClient = rpc.NewClient(c.addr, context)
 		select {
 		case <-c.rpcClient.Healthy():
-			// Success!
+			// Start gossipping and wait for disconnect or error.
```
Comment (spencerkimball) [inline gossip/client.go]: By the way, it's "gossiping", not "gossipping". I grepped the code and it seems there are two instances of "gossipping". Could you clean them all up?

File: gossip/client.go [hunk 3]
```diff
@@ -60,23 +59,29 @@ func newClient(addr net.Addr) *client {
...
+			c.lastFresh = time.Now().UnixNano()
+			err = c.gossip(g, stopper)
+			if context.DisableCache {
+				c.rpcClient.Close()
+			}
 		case <-c.rpcClient.Closed:
-			c.err = util.Errorf("gossip client was closed")
-			done <- c
+			err = util.Error("client closed")
 			return
```
Comment (spencerkimball) [inline gossip/client.go]: Remove this return statement? It's confusing.

File: gossip/client.go [hunk 4]
```diff
@@ -60,23 +59,29 @@ func newClient(addr net.Addr) *client {
 // returns immediately.
 func (c *client) start(g *Gossip, done chan *client, context *rpc.Context, stopper *stop.Stopper) {
 	stopper.RunWorker(func() {
+		var err error
+
+		defer func() {
```
Comment (spencerkimball) [inline gossip/client.go]: Why does this need to be a deferred function?

File: gossip/client.go [hunk 5]
```diff
@@ -94,24 +99,24 @@ func (c *client) gossip(g *Gossip, stopper *stop.Stopper) error {
 	for {
 		// Compute the delta of local node's infostore to send with request.
 		g.mu.Lock()
-		delta := g.is.delta(c.peerID, localMaxSeq)
+		infos := g.is.delta(c.peerID, localMaxSeq)
```
Comment (spencerkimball) [inline gossip/client.go]: I think it's still clearer to call this variable delta.

File: gossip/gossip.proto [hunk 6]
```diff
@@ -33,61 +33,44 @@ option (gogoproto.goproto_unrecognized_all) = false;
...
   // allow gossip to know when client connections are dropped).
-  optional util.UnresolvedAddr l_addr = 3 [(gogoproto.nullable) = false];
+  util.UnresolvedAddr l_addr = 3 [(gogoproto.nullable) = false];
   // Maximum sequence number of gossip from this peer.
-  optional int64 max_seq = 4 [(gogoproto.nullable) = false];
-  // Reciprocal delta of new info since last gossip.
-  optional InfoStoreDelta delta = 5 [(gogoproto.nullable) = false];
+  int64 max_seq = 4;
+  // Delta of new Infos since last gossip.
+  map<string, Info> infos = 5;
```
Comment (spencerkimball) [inline gossip/gossip.proto]: I think you should keep this variable named "delta". It's so much more descriptive.

File: gossip/gossip.proto [hunk 7]
```diff
@@ -33,61 +33,44 @@ option (gogoproto.goproto_unrecognized_all) = false;
...
-  // Maximum sequence number inserted
-  optional int64 max_seq = 4 [(gogoproto.nullable) = false];
+  // Address of the responding client.
+  util.UnresolvedAddr addr = 2 [(gogoproto.nullable) = false];
+  // Non-nil means client should retry with this address.
+  util.UnresolvedAddr alternate = 3;
+  // Maximum sequence number of gossip from this peer.
+  int64 max_seq = 4;
+  // Requested delta of server's infostore.
+  map<string, Info> infos = 5;
```
Comment (spencerkimball) [inline gossip/gossip.proto]: Ditto here.

**Comment by spencerkimball** [comment]
LGTM

File: gossip/client.go (hunk 2 above)
Comment (tamird) [inline gossip/client.go]: done

File: gossip/client.go (hunk 3 above)
Comment (tamird) [inline gossip/client.go]: done

File: gossip/client.go (hunk 4 above)
Comment (tamird) [inline gossip/client.go]: not sure! removed

File: gossip/client.go (hunk 1 above)
Comment (tamird) [inline gossip/client.go]: It's a weird error handling pattern and wasn't actually useful in practice; Right now the error is logged as soon as a client is released, which I think is much more useful for anyone trying to match the error to a line in the source.

File: gossip/client.go (hunk 5 above)
Comment (tamird) [inline gossip/client.go]: done

File: gossip/gossip.proto (hunk 6 above)
Comment (tamird) [inline gossip/gossip.proto]: done

File: gossip/gossip.proto (hunk 7 above)
Comment (tamird) [inline gossip/gossip.proto]: done

**Comment by BramGruneir** [comment]
Besides the one issue, LGTM
